### ISSUE TITLE
Add Pod priorityClassName

### ISSUE THREAD

**Issue by jsalatiel** [kyverno/policies#661]
The  add-pod-priorityclassname sample is broken:

```
    - name: add-priorityclass-pods
      match:
        any:
        - resources:
            kinds:
            - Pod
            namespaceSelector:
              matchExpressions:
                - key: env
                  operator: NotIn
                  values:
                  - production
      mutate:
        patchesJson6902: |-
          - op: remove
            path: '/spec/priority'
          - op: add
            path: /spec/priorityClassName
            value: 'non-production'   
```

The policy is removing the /spec/priority which is actually real the value k8s uses to priorities. Since it is changed at pod level, actually all pods are being internally set as priority=0 ignoring the real value of 'non-production' priority class.

**Comment by chipzoller** [comment]
The reason this was done, as I recall, is because how that internal controller works. When a Pod is created it'll already have that priority set. If it needs to be mutated, the priority field alone is not enough to set. The priority class is used to determine the priority, so you have to remove priority while you add the class in order for the new priority name to be set.

**Comment by jsalatiel** [comment]
What i am trying to say is that if you try to mutate at pod level you are effectively breaking priorities .
The change must be done only on the controllers/cronjob level and it will work.

I noticed the problem when i got a failure on a production cluster.

**Comment by chipzoller** [comment]
@mintbomb27, can you look into this one from a test perspective, please?

**Comment by mintbomb27** [comment]
I think we'll have to try and schedule a good number of pods, and maybe check the age? Or we could check if a low priority pod was evicted from a node where a high priority pod has affinity to(might have to make changes to resource allocations as well). These are if we want to test if the scheduling is happening based on the priority.

About the `priority` key, it is set from the `priorityClassName` automatically if we mention a `priorityClassName` while creating a pod. The policy kind of removes the priority, so it might cause problems as @jsalatiel mentioned.

I checked if it's added back by the priority admission controller, but that doesn't happen. If we don't remove the existing `priority` like how the policy does now, the following error happens:
```
Error from server (Forbidden): error when creating "pod.yaml": pods "blank" is forbidden: the integer value of priority (0) must not be provided in pod spec; priority admission controller computed 1234 from the given PriorityClass name
```

So what we can do, is change the policy to set the priority along with the `priorityClassName` maybe through a context variable fetched from the PriorityClass itself, so that we can avoid this issue.

**Comment by shubham-cmyk** [comment]
Why are we trying to remove the pod priority here? can't we just update the priority class Name since the scheduling is done after the object is persisted to etcd and the mutation would be done before the object is persisted to etcd. 

I maybe wrong here any suggestion @chipzoller @mintbomb27 

```
      mutate:
        patchesJson6902: |-
          - op: remove
            path: '/spec/priority'
          - op: add
            path: /spec/priorityClassName
            value: 'non-production'   
```

**Comment by chipzoller** [comment]
@shubham-cmyk the reason the removal must be done first is because an update of the field is not possible.

@mintbomb27, let's just "fix" this by removing the mutation on the bare Pod and leave it in place for Pod controllers.

**Comment by jsalatiel** [comment]
> @shubham-cmyk the reason the removal must be done first is because an update of the field is not possible.
> 
> @mintbomb27, let's just "fix" this by removing the mutation on the bare Pod and leave it in place for Pod controllers.

That's what i did.
